frost-secp256k1-tr/src/lib.rs

```diff
@@ -0,0 +1,657 @@
+pub fn tweaked_secret_key(
+    secret: <<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Field as Field>::Scalar,
+    public_key: &<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element,
+    merkle_root: &[u8],
+) -> <<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Field as Field>::Scalar {
+    if public_key.to_affine().y_is_odd().into() {
+        -secret + tweak(&public_key, merkle_root)
+    } else {
+        secret + tweak(&public_key, merkle_root)
+    }
+}
+
+impl Ciphersuite for Secp256K1Sha256 {
+    const ID: &'static str = CONTEXT_STRING;
+
+    type Group = Secp256K1Group;
+
+    type HashOutput = [u8; 32];
+
+    type SignatureSerialization = [u8; 65];
+
+    /// H1 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.1
+    fn H1(m: &[u8]) -> <<Self::Group as Group>::Field as Field>::Scalar {
+        hash_to_scalar((CONTEXT_STRING.to_owned() + "rho").as_bytes(), m)
+    }
+
+    /// H2 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.2
+    fn H2(m: &[u8]) -> <<Self::Group as Group>::Field as Field>::Scalar {
+        let mut hasher = tagged_hash("BIP0340/challenge");
+        hasher.update(m);
+        hasher_to_scalar(hasher)
+    }
+
+    /// H3 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.3
+    fn H3(m: &[u8]) -> <<Self::Group as Group>::Field as Field>::Scalar {
+        hash_to_scalar((CONTEXT_STRING.to_owned() + "nonce").as_bytes(), m)
+    }
+
+    /// H4 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.4
+    fn H4(m: &[u8]) -> Self::HashOutput {
+        hash_to_array(&[CONTEXT_STRING.as_bytes(), b"msg", m])
+    }
+
+    /// H5 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.5
+    fn H5(m: &[u8]) -> Self::HashOutput {
+        hash_to_array(&[CONTEXT_STRING.as_bytes(), b"com", m])
+    }
+
+    /// HDKG for FROST(secp256k1, SHA-256)
+    fn HDKG(m: &[u8]) -> Option<<<Self::Group as Group>::Field as Field>::Scalar> {
+        Some(hash_to_scalar(
+            (CONTEXT_STRING.to_owned() + "dkg").as_bytes(),
+            m,
+        ))
+    }
+
+    /// HID for FROST(secp256k1, SHA-256)
+    fn HID(m: &[u8]) -> Option<<<Self::Group as Group>::Field as Field>::Scalar> {
+        Some(hash_to_scalar(
+            (CONTEXT_STRING.to_owned() + "id").as_bytes(),
+            m,
+        ))
+    }
+
+    /// Generates the challenge as is required for Schnorr signatures.
+    fn challenge(R: &Element<S>, verifying_key: &VerifyingKey, msg: &[u8]) -> Challenge<S> {
+        let mut preimage = vec![];
+        let tweaked_public_key = tweaked_public_key(&verifying_key.to_element(), &[]);
```

Comment:
Should the merkel root be empty here?
i.e does this code assume that we only spend from the default key path? No a tapscript path?

Reply:
Possibly my understanding is wrong, but in BIP341, in
code example for `taproot_sign_key` I've seen next:
```
    if script_tree is None:
        h = bytes()  # empty bytes for merkle root
```

In the case when script path is not used (used only key path),
seems empty bytes for merkle root is correct.

Reply:
yeah this code only works with empty commitment, and it works fine for us (https://github.com/sigma0-xyz/zkbitcoin). I think it's a good first step, but it'd be nice to expose that feature later

Reply:
Thank you. I'll look to code in zkbitcoin in a few days.

Reply:
>In the case when script path is not used (used only key path),

Ah yes you're correct. In the case of a script path spend (non-frost spend) the spender can tweak the verifying key outside the context of this library.

Reply:
> Thank you. I'll look to code in zkbitcoin in a few days.

Only looking in the current day...